mdapi/MetricsDiscoveryHelper.cpp

```diff
@@ -0,0 +1,1142 @@
+/* ActivateMetricSet                                                    */
+/************************************************************************/
+void MDHelper::ActivateMetricSet()
+{
+    if( !m_Initialized || !m_MetricSet ) return;
+
+    TCompletionCode res = m_MetricSet->Activate();
+    if( res != CC_OK ) DebugPrint("ActivateMetricSet failed!\n");
+}
+
+/************************************************************************/
+/* DeactivateMetricSet                                                  */
+/************************************************************************/
+void MDHelper::DeactivateMetricSet()
+{
+    if( !m_Initialized || !m_MetricSet ) return;
+
+    TCompletionCode res = m_MetricSet->Deactivate();
+    if( res != CC_OK ) DebugPrint("DeactivateMetricSet failed!\n");
+}
+
+/************************************************************************/
+/* SetMetricSetFiltering                                                */
+/************************************************************************/
+void MDHelper::SetMetricSetFiltering( TMetricApiType apiMask )
+{
+    if( !m_Initialized || !m_MetricSet ) return;
+
+    TCompletionCode res = m_MetricSet->SetApiFiltering( apiMask );
+    if( res != CC_OK ) DebugPrint("SetMetricSetFiltering failed!\n");
+}
+
+/************************************************************************/
+/* GetMetricsFromReport                                                 */
+/************************************************************************/
+void MDHelper::GetMetricsFromReport(
+    const char* pReportData,
+    std::vector<TTypedValue_1_0>& results,
+    std::vector<TTypedValue_1_0>& information )
+{
+    if( !m_Initialized || !m_MetricSet ) return;
+
+    const uint32_t metricsCount     = m_MetricSet->GetParams()->MetricsCount;
+    const uint32_t informationCount = m_MetricSet->GetParams()->InformationCount;
+
+    std::vector<TTypedValue_1_0>    deltaValues;
+
+    deltaValues.resize( metricsCount );
+    results.resize( metricsCount );
+    information.resize( informationCount );
+
+    // Read metrics from the report data and normalize:
+    ReadMetrics( pReportData, deltaValues );
+    NormalizeMetrics( deltaValues, results );
+
+    // Read informational from the report data:
+    ReadInformation( pReportData, information );
+}
+
+/************************************************************************/
+/* PrintMetricNames                                                     */
+/************************************************************************/
+void MDHelper::PrintMetricNames( std::ostream& os )
+{
+    if( !m_Initialized || !m_MetricSet || !os.good() ) return;
+
+    os << "kernel,";
+
+    for( uint32_t i = 0; i < m_MetricSet->GetParams( )->MetricsCount; i++ )
+    {
+        // Skip if not supported
+        if ( 0 == ( m_MetricSet->GetMetric( i )->GetParams()->ApiMask & m_APIMask )) continue;
```

Comment:
Checking ApiMask isn't necessary if you use SetApiFiltering (which is btw required for using CalculateMetrics()), please remove (same for other places). SetApiFiltering filters out all unavailable metrics on the set API.